From 2e0cacc7426c64c725f71368a51a3f00a821d3ee Mon Sep 17 00:00:00 2001
From: Pau Ruiz Safont <pau.safont@vates.tech>
Date: Mon, 9 Feb 2026 10:31:47 +0100
Subject: [PATCH] xapi_sr_operations: ensure the properties on ops lists always
 hold

The lists of operations used in the module need to be a subset of
all_ops, otherwise spurious errors can be thrown on runtime.

Ensuring these properties allows to drop the runtime check

Also add an interface file for the module

Signed-off-by: Pau Ruiz Safont <pau.safont@vates.tech>
---
 ocaml/tests/dune                        | 14 +++++---
 ocaml/tests/test_xapi_sr_operations.ml  | 44 +++++++++++++++++++++++++
 ocaml/tests/test_xapi_sr_operations.mli |  0
 ocaml/xapi/xapi_sr_operations.ml        |  6 +---
 ocaml/xapi/xapi_sr_operations.mli       | 40 ++++++++++++++++++++++
 quality-gate.sh                         |  2 +-
 6 files changed, 96 insertions(+), 10 deletions(-)
 create mode 100644 ocaml/tests/test_xapi_sr_operations.ml
 create mode 100644 ocaml/tests/test_xapi_sr_operations.mli
 create mode 100644 ocaml/xapi/xapi_sr_operations.mli

diff --git a/ocaml/tests/dune b/ocaml/tests/dune
index 2f51f4f12..4908fd193 100644
--- a/ocaml/tests/dune
+++ b/ocaml/tests/dune
@@ -132,12 +132,18 @@
   (preprocess (per_module ((pps ppx_deriving_rpc) Test_cluster_host)))
 )
 
-(test
-(name test_storage_smapiv1_wrapper)
+(tests
+(names test_storage_smapiv1_wrapper test_xapi_sr_operations)
 (modes exe)
 (package xapi)
-(modules test_storage_smapiv1_wrapper)
-(libraries alcotest xapi_internal fmt xapi-idl.storage.interface xapi-idl.storage.interface.types))
+(modules test_storage_smapiv1_wrapper test_xapi_sr_operations)
+(libraries
+  alcotest
+  fmt
+  record_util
+  xapi-idl.storage.interface
+  xapi-idl.storage.interface.types
+  xapi_internal))
 
 (test
 (name test_storage_quicktest)
diff --git a/ocaml/tests/test_xapi_sr_operations.ml b/ocaml/tests/test_xapi_sr_operations.ml
new file mode 100644
index 000000000..5045c5e8f
--- /dev/null
+++ b/ocaml/tests/test_xapi_sr_operations.ml
@@ -0,0 +1,44 @@
+(*
+  Copyright (C) 2026 Vates.
+ 
+  This program is free software; you can redistribute it and/or modify
+  it under the terms of the GNU Lesser General Public License as published
+  by the Free Software Foundation; version 2.1 only. with the special
+  exception on linking described in file LICENSE.
+ 
+  This program is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU Lesser General Public License for more details.
+ *)
+
+module Ops = Xapi_sr_operations
+
+let contained_tables =
+  [
+    (Ops.all_rpu_ops, "all_rpu_ops")
+  ; (Ops.disallowed_during_rpu, "disallowed_during_rpu")
+  ; (Ops.sm_cap_table |> List.map fst, "sm_cap_table")
+  ]
+
+let not_contained element =
+  if not (List.mem element Ops.all_ops) then
+    Some (Record_util.storage_operations_to_string element)
+  else
+    None
+
+let test_tables =
+  List.map
+    (fun (input, name) ->
+      let test () =
+        let not_contained = List.filter_map not_contained input in
+        Alcotest.(check @@ list string)
+          "There cannot be operations missing from all_ops" [] not_contained
+      in
+      (Printf.sprintf {|%s is a subset of all_ops|} name, `Quick, test)
+    )
+    contained_tables
+
+let () =
+  Alcotest.run "Test XAPI Helpers suite"
+    [("SR operation tables all are complete ", test_tables)]
diff --git a/ocaml/tests/test_xapi_sr_operations.mli b/ocaml/tests/test_xapi_sr_operations.mli
new file mode 100644
index 000000000..e69de29bb
diff --git a/ocaml/xapi/xapi_sr_operations.ml b/ocaml/xapi/xapi_sr_operations.ml
index a9f3bf7eb..e74e19863 100644
--- a/ocaml/xapi/xapi_sr_operations.ml
+++ b/ocaml/xapi/xapi_sr_operations.ml
@@ -271,13 +271,9 @@ let valid_operations ~__context ?op record _ref' : table =
 
 let throw_error (table : table) op =
   match Hashtbl.find_opt table op with
-  | None ->
-      Helpers.internal_error
-        "xapi_sr.assert_operation_valid unknown operation: %s"
-        (sr_operation_to_string op)
   | Some (Error ex) ->
       raise ex
-  | Some (Ok ()) ->
+  | None | Some (Ok ()) ->
       ()
 
 let assert_operation_valid ~__context ~self ~(op : API.storage_operations) =
diff --git a/ocaml/xapi/xapi_sr_operations.mli b/ocaml/xapi/xapi_sr_operations.mli
new file mode 100644
index 000000000..e4fce0189
--- /dev/null
+++ b/ocaml/xapi/xapi_sr_operations.mli
@@ -0,0 +1,40 @@
+type table = (API.storage_operations, (unit, exn) Result.t) Hashtbl.t
+
+val features_of_sr_internal :
+  __context:Context.t -> _type:string -> (Smint.Feature.capability * int64) list
+
+val features_of_sr :
+     __context:Context.t
+  -> Db_actions.sR_t
+  -> (Smint.Feature.capability * int64) list
+
+val assert_operation_valid :
+     __context:Context.t
+  -> self:[`SR] API.Ref.t
+  -> op:API.storage_operations
+  -> unit
+
+val update_allowed_operations :
+  __context:Context.t -> self:[`SR] API.Ref.t -> unit
+
+val cancel_tasks :
+     __context:Context.t
+  -> self:[`SR] API.Ref.t
+  -> all_tasks_in_db:[`task] Ref.t list
+  -> task_ids:string list
+  -> unit
+
+val sr_health_check : __context:Context.t -> self:[`SR] API.Ref.t -> unit
+
+val stop_health_check_thread :
+  __context:Context.t -> self:[`SR] API.Ref.t -> unit
+
+(**/**)
+
+val all_ops : API.storage_operations_set
+
+val all_rpu_ops : API.storage_operations_set
+
+val disallowed_during_rpu : API.storage_operations_set
+
+val sm_cap_table : (API.storage_operations * Smint.Feature.capability) list
diff --git a/quality-gate.sh b/quality-gate.sh
index 54fd5adc4..c1df9d96f 100755
--- a/quality-gate.sh
+++ b/quality-gate.sh
@@ -25,7 +25,7 @@ verify-cert () {
 }
 
 mli-files () {
-  N=456
+  N=455
   X="ocaml/tests"
   X+="|ocaml/quicktest"
   X+="|ocaml/message-switch/core_test"
